keymanager/src/crypto/kdf.rs

```diff
@@ -674,46 +801,92 @@ impl Kdf {
     }
 
-    /// Load the generation of the last stored master secret from untrusted local storage.
-    fn load_last_generation(untrusted_local: &dyn KeyValue, runtime_id: &Namespace) -> Option<u64> {
-        // Fetch the encrypted generation if it exists.
-        let key = LATEST_GENERATION_STORAGE_KEY.to_vec();
-        let ciphertext = untrusted_local.get(key).unwrap();
+    /// Load the proposal for the next master secret from untrusted local storage.
+    ///
+    /// Since master secret proposals can be overwritten if not accepted by the end of the rotation
+    /// period, it is impossible to know whether the loaded proposal is the latest one. Therefore,
+    /// it is crucial to ALWAYS verify that the checksum of the proposal matches the one published
+    /// in the consensus before accepting it.
+    fn load_master_secret_proposal(storage: &dyn KeyValue) -> Option<Secret> {
+        // Fetch the encrypted master secret proposal if it exists.
+        let key = MASTER_SECRET_PROPOSAL_STORAGE_KEY.to_vec();
+
+        let ciphertext = storage.get(key).unwrap();
         if ciphertext.is_empty() {
             return None;
         }
 
-        let (ciphertext, nonce) = unpack_encrypted_generation_nonce(&ciphertext)
+        let (ciphertext, nonce) = unpack_encrypted_secret_nonce(&ciphertext)
             .expect("persisted state is corrupted, invalid size");
 
-        // Decrypt the persisted generation.
+        // Decrypt the persisted master secret proposal.
         let d2 = Self::new_d2();
-        let plaintext = d2
-            .open(&nonce, ciphertext.to_vec(), runtime_id)
-            .expect("persisted state is corrupted");
+        let plaintext = match d2.open(&nonce, ciphertext.to_vec(), vec![]) {
+            Ok(plaintext) => plaintext,
+            Err(_) => return None,
+        };
 
-        Some(u64::from_le_bytes(plaintext.try_into().unwrap()))
+        Some(Secret(plaintext.try_into().unwrap()))
     }
 
-    /// Store the generation of the last master secret to untrusted local storage.
-    fn store_last_generation(
-        untrusted_local: &dyn KeyValue,
-        runtime_id: &Namespace,
-        generation: u64,
-    ) {
-        // We only store the latest generation.
-        let key = LATEST_GENERATION_STORAGE_KEY.to_vec();
+    /// Encrypt and store the next master secret proposal in untrusted local storage.
+    ///
+    /// If a proposal already exists, it will be overwritten.
+    fn store_master_secret_proposal(storage: &dyn KeyValue, secret: &Secret) {
+        // Using the same key for all proposals will override the previous one.
+        let key = MASTER_SECRET_PROPOSAL_STORAGE_KEY.to_vec();
 
-        // Encrypt the generation.
+        // Encrypt the master secret.
+        // Additional data has to be different from the one used when storing verified master
+        // secrets so that the attacker cannot replace secrets with rejected proposals.
```

Comment:
Ah, we should add explicit domain separation contexts to `new_d2` similar to how we have it in the `common/sgx/seal` implementation. Currently everything here is just using `MASTER_SECRET_SEAL_CONTEXT` (because previously there were no other data to seal) which should be avoided.

We should also change the seal contexts to follow the new context scheme (they depend on MRENCLAVE anyway so no backward compatibility is needed).

Reply:
could make a more general context and use the additional data to ensure sealed stuff isn't misused

Reply:
> could make a more general context and use the additional data to ensure sealed stuff isn't misused

Yeah this is how it is currently done in this PR, but it seems less explicit.

Reply:
Fixed, added/renamed contexts for master secrets and master secret proposals.